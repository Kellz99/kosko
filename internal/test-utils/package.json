{
  "name": "@kosko/test-utils",
  "version": "0.0.0",
  "private": true,
  "engines": {
    "node": ">=14.18.0"
  },
  "main": "dist/index.node.cjs",
  "module": "dist/index.node.mjs",
  "types": "dist/index.d.ts",
  "sideEffects": false,
  "scripts": {
    "build": "build-utils"
  },
  "exports": {
    "node": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.node.mjs",
      "require": "./dist/index.node.cjs"
    },
    "types": "./dist/index.d.ts",
    "import": "./dist/index.base.mjs",
    "require": "./dist/index.node.cjs"
  },
  "dependencies": {
    "execa": "^5.1.1",
    "temp-dir": "^2.0.0",
    "tmp": "^0.2.1",
    "tmp-promise": "^3.0.3"
  },
  "devDependencies": {
    "@kosko/build-utils": "workspace:^"
  }
}
