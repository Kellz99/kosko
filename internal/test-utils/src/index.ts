export * from "./installPackage";
export * from "./projectRoot";
export * from "./runCLI";
export { sleep, getRejectedValue } from "./promise";
export { makeTempDir, makeTempFile, type TempDir, type TempFile } from "./temp";
