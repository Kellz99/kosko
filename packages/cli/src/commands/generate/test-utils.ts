import { Manifest } from "@kosko/generate";

export function createManifest(
  value: Partial<Omit<Manifest, "report">> = {}
): Manifest {
  return {
    position: { path: "", index: [] },
    issues: [],
    data: undefined,
    report() {},
    ...value
  };
}
