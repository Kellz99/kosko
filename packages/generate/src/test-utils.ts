/// <reference types="jest-extended" />
import { Manifest } from "./base";

type ManifestToMatch = Pick<Manifest, "data" | "metadata"> &
  Partial<Pick<Manifest, "position" | "issues">>;

export function matchManifest({
  position = { path: "", index: [] },
  issues = [],
  ...manifest
}: ManifestToMatch) {
  return {
    ...manifest,
    position,
    issues,
    report: expect.toBeFunction()
  };
}

export function matchManifests(manifests: ManifestToMatch[]) {
  return manifests.map(matchManifest);
}
