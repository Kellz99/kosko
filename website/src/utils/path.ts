export function isAbsolute(path: string): boolean {
  return path.startsWith("/");
}

export function isRelative(path: string): boolean {
  return path.startsWith("./") || path.startsWith("../");
}

export function dirname(path: string): string {
  const index = path.lastIndexOf("/");
  if (index === -1) return path;
  return path.substring(0, index - 1) || "/";
}

export function extname(path: string): string {
  const index = path.lastIndexOf(".");
  if (index === -1) return "";
  return path.substring(index);
}

export function basename(path: string, ext?: string): string {
  const index = path.lastIndexOf("/");
  const base = index === -1 ? path : path.substring(index + 1);

  if (base.endsWith(ext)) {
    return base.substring(0, base.length - ext.length);
  }

  return base;
}

export function relative(from: string, to: string): string {
  const fromParts = from.split("/").filter(Boolean);
  const toParts = to.split("/").filter(Boolean);

  while (fromParts[0] && toParts[0]) {
    fromParts.shift();
    toParts.shift();
  }

  while (toParts[0] === "." || toParts[0] === "..") {
    const toPart = toParts.shift();

    if (toPart === "..") {
      fromParts.pop();
    }
  }

  while (fromParts.pop()) {
    toParts.unshift("..");
  }

  return toParts.join("/");
}
