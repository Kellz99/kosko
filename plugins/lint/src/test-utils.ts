import { Issue, Manifest, type ManifestToValidate } from "@kosko/generate";
import type { RuleContext, RuleFactory } from "./rules/types";
import assert from "node:assert";

export interface ReportedIssue {
  manifest: ManifestToValidate;
  message: string;
}

export interface TestContext<T> extends RuleContext<T> {
  issues: ReportedIssue[];
}

export function createTestContext<T>(config: T): TestContext<T> {
  const issues: ReportedIssue[] = [];

  return {
    config,
    severity: "error",
    issues,
    report(manifest, message) {
      issues.push({ manifest, message });
    }
  };
}

export function createManifest(
  value: unknown
): ManifestToValidate & Pick<Manifest, "issues"> {
  const issues: Issue[] = [];

  return {
    path: "",
    index: [],
    data: value,
    issues,
    report(issue) {
      issues.push(issue);
    }
  };
}

export function validate<T>(
  rule: RuleFactory<T>,
  config: T,
  manifest: ManifestToValidate
): ReportedIssue[] {
  const ctx = createTestContext(config);
  const { validate } = rule.factory(ctx);

  assert(typeof validate === "function");
  validate(manifest);

  return ctx.issues;
}

export function validateAll<T>(
  rule: RuleFactory<T>,
  config: T,
  manifests: readonly ManifestToValidate[]
): ReportedIssue[] {
  const ctx = createTestContext(config);
  const { validateAll } = rule.factory(ctx);

  assert(typeof validateAll === "function");
  validateAll(manifests);

  return ctx.issues;
}
