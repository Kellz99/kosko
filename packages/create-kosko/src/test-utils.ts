import { join, posix } from "node:path";
import glob from "fast-glob";
import { readFile } from "node:fs/promises";

export async function listAllFiles(
  dir: string
): Promise<Record<string, string>> {
  const paths = await glob("**/*", { cwd: dir });
  const files: Record<string, string> = {};

  for (const path of paths) {
    files[posix.normalize(path)] = await readFile(join(dir, path), "utf8");
  }

  return files;
}
