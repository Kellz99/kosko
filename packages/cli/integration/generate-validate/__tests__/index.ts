/// <reference types="jest-extended"/>
import execa from "execa";
import { dirname } from "path";
import { runCLI, installPackage } from "@kosko/test-utils";

const testDir = dirname(__dirname);

let args: string[];
let result: execa.ExecaReturnValue;
let options: execa.Options;

beforeAll(async () => {
  await installPackage(testDir, "env");
});

beforeEach(async () => {
  result = await runCLI(args, {
    ...options,
    cwd: testDir
  });
});

describe("when validate is not set", () => {
  beforeAll(() => {
    args = ["generate"];
    options = { reject: false };
  });

  test("should return status code 1", () => {
    expect(result.exitCode).toEqual(1);
  });

  test("should print the error", () => {
    expect(result.stderr).toContain(
      "Error: data/spec must have required property 'containers', data/spec must be null, data/spec must match exactly one schema in oneOf"
    );
  });
});

describe("when validate is true", () => {
  beforeAll(() => {
    args = ["generate", "--validate"];
    options = { reject: false };
  });

  test("should return status code 1", () => {
    expect(result.exitCode).toEqual(1);
  });

  test("should print the error", () => {
    expect(result.stderr).toContain(
      "Error: data/spec must have required property 'containers', data/spec must be null, data/spec must match exactly one schema in oneOf"
    );
  });
});

describe("when validate is false", () => {
  beforeAll(() => {
    args = ["generate", "--validate", "false"];
    options = {};
  });

  test("should return status code 0", () => {
    expect(result.exitCode).toEqual(0);
  });

  test("should print the manifest", () => {
    expect(result.stdout).toMatchSnapshot();
  });
});
