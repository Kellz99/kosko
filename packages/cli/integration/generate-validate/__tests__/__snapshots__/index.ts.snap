// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when validate is false should print the manifest 1`] = `
"apiVersion: v1
kind: Pod
metadata:
  name: foo
spec: {}"
`;

exports[`when validate is not set should print the error 1`] = `
"
[1mcomponents/foo.js[22m - [31mâœ– 1[39m

  Kind: v1/Pod, Name: foo
    [31mâœ–[39m /spec must have required property 'containers'

[31merror[39m - Generate failed (Total [31mâœ– 1[39m)"
`;

exports[`when validate is true should print the error 1`] = `
"
[1mcomponents/foo.js[22m - [31mâœ– 1[39m

  Kind: v1/Pod, Name: foo
    [31mâœ–[39m /spec must have required property 'containers'

[31merror[39m - Generate failed (Total [31mâœ– 1[39m)"
`;
